[TensorExpr] Fix aten::atan2 lowering and disable aten::pow lowering on CPU.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48355 [TensorExpr] Add a envvar to disable LLVM backend and use IR Eval instead.
* #48354 [TensorExpr] IREval: fix div for Half dtype.
* #48329 [TensorExpr] Fix lowering for aten::div.
* **#48326 [TensorExpr] Fix aten::atan2 lowering and disable aten::pow lowering on CPU.**

The PR introduces a set of 'cuda-only' ops into `isSupported` function.
It is done to disable `pow` lowering on CPU where it's tricky to support
integer versions.

Differential Revision: [D25129211](https://our.internmc.facebook.com/intern/diff/D25129211)

Commit:
[TensorExpr] Fix aten::atan2 lowering and disable aten::pow lowering on CPU.

The PR introduces a set of 'cuda-only' ops into `isSupported` function.
It is done to disable `pow` lowering on CPU where it's tricky to support
integer versions.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -169,6 +172,15 @@ bool isSupported(Node* node) {
+    if (node->isMemberOf(cuda_only_operator_set)) {
+      auto device = tensorexpr::pickDeviceType(node->inputs());
+      if (!device) {
+        device = tensorexpr::pickDeviceType(node->outputs());
```
I think we require that the types be fully known for inputs & outputs, so in theory this conditional shouldn't matter.

ZolotukhinM replied:
The reason I put it there is that maybe we could have an op with 0 tensor inputs - in this case we won't be able to deduce device from there. I don't think it could happen now, but better safe than sorry :)

eellison replied:
Oh, the other thing is that we're no longer checking that `exponent` has a known int / float type, which we need to know to compile it

ZolotukhinM replied:
I'm not sure I follow: this change doesn't relax existing checks, only adds a new device check. So, if we don't check it now, we didn't check it before.

Also, this PR adds a test that iterates over all dtypes for the cases we do fuse - and it seems that on cuda the `pow.Tensor_Scalar` works correctly for all dtypes except float16 (which we should fix separately).

eellison replied:
yep, brain fryed from being the oncall 🤡  you're right we're still doing those checks

eellison commented on test/test_tensorexpr.py:
```diff
@@ -967,7 +967,7 @@ def test_min(x, y):
 
     def test_double_intrinsics(self):
-        devices = ["cuda", "cpu"] if torch.cuda.is_available() else ["cpu"]
+        devices = ["cuda"] if torch.cuda.is_available() else []
```
coudl you add a TODO here ?

Commit:
Update on "[TensorExpr] Fix aten::atan2 lowering and disable aten::pow lowering on CPU."


The PR introduces a set of 'cuda-only' ops into `isSupported` function.
It is done to disable `pow` lowering on CPU where it's tricky to support
integer versions.

Differential Revision: [D25129211](https://our.internmc.facebook.com/intern/diff/D25129211)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fix aten::atan2 lowering and disable aten::pow lowering on CPU."


The PR introduces a set of 'cuda-only' ops into `isSupported` function.
It is done to disable `pow` lowering on CPU where it's tricky to support
integer versions.

Differential Revision: [D25129211](https://our.internmc.facebook.com/intern/diff/D25129211)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Fix aten::atan2 lowering and disable aten::pow lowering on CPU."


The PR introduces a set of 'cuda-only' ops into `isSupported` function.
It is done to disable `pow` lowering on CPU where it's tricky to support
integer versions.

Differential Revision: [D25129211](https://our.internmc.facebook.com/intern/diff/D25129211)

[ghstack-poisoned]

eellison reviewed (commented):
What is the issue with pow ? just we don't support it well for integers... Should we re-enable with only supporting float?
